src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -8,108 +8,208 @@
+            }
+
+            Chunk chunk = generateChunk(block_size - result.getNumRows());
+
+            if (result)
+                result.append(chunk);
+            else
+                result = std::move(chunk);
         }
+
+        progress(result.getNumRows(), result.bytes());
+        return result;
     }
 
-    MutableColumns new_columns = header.cloneEmptyColumns();
-    while (StoragesInfo info = stream.next())
+private:
+    std::shared_ptr<SourceState> state;
+    const std::vector<UInt8> columns_mask;
+    const UInt64 block_size;
+    const bool has_bytes_on_disk_column;
+    static const size_t files_peer_thread = 15;
+
+    StoragesInfo current_info;
+    DetachedPartsInfo detached_parts;
+
+    void getMoreParts()
     {
-        const auto parts = info.data->getDetachedParts();
-        for (const auto & p : parts)
+        chassert(detached_parts.empty());
+
+        current_info = state->next();
+        if (!current_info)
+            return;
+
+        detached_parts = current_info.data->getDetachedParts();
+    }
+
+    void calculatePartSizeOnDisk(size_t begin, std::vector<std::atomic<size_t>> & parts_sizes)
+    {
+        if (!has_bytes_on_disk_column)
+            return;
+
+        std::vector<std::future<void>> futures;
+        SCOPE_EXIT_SAFE({
+            /// Exceptions are not propagated
+            for (auto & future : futures)
+                if (future.valid())
+                    future.wait();
+            futures.clear();
+        });
+
+        for (auto p_id = begin; p_id < detached_parts.size(); ++p_id)
         {
-            size_t src_index = 0, res_index = 0;
+            auto & p = detached_parts.at(p_id);
+            DiskPtr & disk = p.disk;
+
+            const String part_path = fs::path(MergeTreeData::DETACHED_DIR_NAME) / p.dir_name;
+            const String relative_path = fs::path(current_info.data->getRelativeDataPath()) / part_path;
+            std::vector<String> listing = partFilesOnDisk(disk, relative_path);
+
+            auto * counter = &parts_sizes[p_id - begin];
+            for (size_t slice_begin = 0; slice_begin < listing.size(); slice_begin += files_peer_thread)
+            {
+                size_t slice_end = std::min(slice_begin + files_peer_thread, listing.size());
+                auto slice = std::vector<String>(listing.begin() + slice_begin, listing.begin() + slice_end);
+                futures.push_back(
+                    scheduleFromThreadPool<void>(
+                        [disk, counter, files = std::move(slice)] ()
+                        {
+                            size_t partial_files_size = 0;
+                            for (const auto & file : files)
+                                partial_files_size += disk->getFileSize(file);
+                            counter->fetch_add(partial_files_size);
+                        },
+                        IOThreadPool::get(),
+                        "DP_BytesOnDisk"));
```

Comment:
Here I afraid a bit that tasks are too small, and a lot of time could still be spend on scheduling.

I would prefer to have a fixed number of threads (even hardcoded constant, like 100), but every thread can increment atomic counter and potentially process much more tasks.

Reply:
Yeah, will try that.

Reply:
With some experiment, I have chosen the value 35. On my data set larger value do not show any profit.